### PR TITLE
Alternate Scoring

### PR DESCRIPTION
**Pull request by insipx**
WIP for #9118

### CODE REVIEW

File: miner/src/pool/scoring.rs [hunk 1]
```diff
@@ -108,6 +204,16 @@ impl<P> txpool::Scoring<P> for NonceAndGasPrice where P: ScoredTransaction + txp
 					super::Priority::Regular => 0,
 				};
 				scores[i] = scores[i] << boost;
+				let last_index = txs.len() - 1;
+				for idx in 1..last_index {
```
Comment (tomusdrw) [inline miner/src/pool/scoring.rs]: Note the score depends on the next score (`score[idx]`) so we should compute it starting from the end

File: miner/src/pool/scoring.rs [hunk 2]
```diff
@@ -108,6 +204,16 @@ impl<P> txpool::Scoring<P> for NonceAndGasPrice where P: ScoredTransaction + txp
...
+					scores[idx - 1] = *txs[idx-1].transaction.gas_price()
+						+ is_consecutive(&txs[idx-1..=idx])
+						+ (U256::from(21_000) / scores[idx - 1])
```
Comment (tomusdrw) [inline miner/src/pool/scoring.rs]: I suppose this should be `gas` not `score`?

File: parity/cli/mod.rs [hunk 3]
```diff
@@ -1346,6 +1347,7 @@ struct Mining {
 	refuse_service_transactions: Option<bool>,
 	infinite_pending_block: Option<bool>,
 	max_round_blocks_to_import: Option<usize>,
+	experimental_scoring: Option<bool>,
```
Comment (tomusdrw) [inline parity/cli/mod.rs]: not needed?

File: parity/helpers.rs [hunk 4]
```diff
@@ -100,6 +100,7 @@ pub fn to_pending_set(s: &str) -> Result<PendingSet, String> {
 pub fn to_queue_strategy(s: &str) -> Result<PrioritizationStrategy, String> {
 	match s {
 		"gas_price" => Ok(PrioritizationStrategy::GasPriceOnly),
+                "consecutive" => Ok(PrioritizationStrategy::Consecutive),
```
Comment (tomusdrw) [inline parity/helpers.rs]: Indentation

File: miner/src/pool/scoring.rs (hunk 2 above)
Comment (insipx) [reply miner/src/pool/scoring.rs]: Yeahp, just didn't add gas() to `ScoredTransaction` yet, so used score as placeholder

File: miner/src/pool/scoring.rs [hunk 5]
```diff
@@ -67,6 +72,7 @@ impl NonceAndGasPrice {
 	}
 }
 
+// differs in the `update_scores` function
```
[the remark this answers is not in the document]
Comment (insipx) [reply miner/src/pool/scoring.rs]: Good point

I've combined them via the PrioritizationStrategy enum for now. However, I think it makes the update_scores function a bit too large imo, so will probably separate the two strategies out once i'm done testing the algorithm for consecutive transactions

**Comment by 5chdn** [comment]
Restarted CI

File: miner/src/pool/scoring.rs [hunk 6]
```diff
@@ -92,33 +92,69 @@ impl<P> txpool::Scoring<P> for NonceAndGasPrice where P: ScoredTransaction + txp
...
+
+		let boost = |i: usize| {
+			match txs[i].priority() {
+				super::Priority::Local => 15,
+				super::Priority::Retracted => 10,
+				super::Priority::Regular => 0,
+			}
+		};
+
+		let event = |scores: &mut [U256]| {
```
Comment (tomusdrw) [inline miner/src/pool/scoring.rs]: maybe call this `penalize`, it took me few seconds to figure out what this function is about (explanation comment on what `event` means is below)

File: miner/src/pool/scoring.rs [hunk 7]
```diff
@@ -92,33 +92,69 @@ impl<P> txpool::Scoring<P> for NonceAndGasPrice where P: ScoredTransaction + txp
...
+				let is_consecutive = |i: usize| {
+					if (txs[i].nonce() + U256::one()) == txs[i+1].nonce() {
+						U256::one()
+					} else {
+						U256::zero()
 					}
+				};
+				match change {
+					Change::Culled(_) => {},
+					Change::RemovedAt(_) => {}
```
Comment (tomusdrw) [inline miner/src/pool/scoring.rs]: So actually it's fine to ignore those events in `GasPriceOnly`, but with `Consecutive` strategy we should recalculate as well.

File: miner/src/pool/scoring.rs [hunk 8]
```diff
@@ -92,33 +92,69 @@ impl<P> txpool::Scoring<P> for NonceAndGasPrice where P: ScoredTransaction + txp
...
+					Change::InsertedAt(i) | Change::ReplacedAt(i) => {
+						assert!(i < txs.len());
+						assert!(i < scores.len());
+						let last_index = txs.len() - 1;
+						scores[last_index] = *txs[last_index].transaction.gas_price();
+						for idx in (1..txs.len()).rev() {
+							let consecutive_bump = is_consecutive(idx - 1) *
```
Comment (tomusdrw) [inline miner/src/pool/scoring.rs]: `idx - 1` is quite common, maybe do `let prev_idx = idx - 1`?

File: miner/src/pool/scoring.rs [hunk 9]
```diff
@@ -92,33 +92,69 @@ impl<P> txpool::Scoring<P> for NonceAndGasPrice where P: ScoredTransaction + txp
...
+								((U256::from(21_000) * scores[idx]) /
+								txs[idx - 1].transaction.gas());
+							scores[idx - 1] = txs[idx - 1].transaction.gas_price() + (consecutive_bump / 1000);
+							scores[idx] <<= boost(idx - 1);
+						}
+						scores[0] <<= boost(last_index);
```
Comment (tomusdrw) [inline miner/src/pool/scoring.rs]: I don't think this is necessary, we already cover that in case `idx == 1`.

File: miner/src/pool/scoring.rs [hunk 10]
```diff
@@ -92,33 +92,69 @@ impl<P> txpool::Scoring<P> for NonceAndGasPrice where P: ScoredTransaction + txp
...
 					}
+				};
+				match change {
+					Change::Culled(_) => {},
+					Change::RemovedAt(_) => {}
+					Change::InsertedAt(i) | Change::ReplacedAt(i) => {
+						assert!(i < txs.len());
+						assert!(i < scores.len());
+						let last_index = txs.len() - 1;
+						scores[last_index] = *txs[last_index].transaction.gas_price();
```
Comment (tomusdrw) [inline miner/src/pool/scoring.rs]: I think you should include `boost` here

Comment (tomusdrw) [reply miner/src/pool/scoring.rs]: Actually, currently `boost` seems to have exponential effect, can you instead apply it either:
1. Doing an additional loop after you calculate base scores (that will boost the scores, but boosting `idx+1` won't affect `idx`)
2. Instead of using `gas_price` use `gas_price` with boost applied.

I'm more inclined to option (2). Would be good to plot some graphs how they differ.

File: parity/cli/mod.rs (hunk 3 above)
Comment (tomusdrw) [reply parity/cli/mod.rs]: I think it's still not needed:)

File: parity/cli/mod.rs [hunk 11]
```diff
@@ -761,9 +761,10 @@ usage! {
 			"--tx-queue-per-sender=[LIMIT]",
 			"Maximum number of transactions per sender in the queue. By default it's 1% of the entire queue, but not less than 16.",
 
+			// INSI: Not sure about the name for this strategy -- 'dynamic'?
 			ARG arg_tx_queue_strategy: (String) = "gas_price", or |c: &Config| c.mining.as_ref()?.tx_queue_strategy.clone(),
 			"--tx-queue-strategy=[S]",
-			"Prioritization strategy used to order transactions in the queue. S may be: gas_price - Prioritize txs with high gas price",
+			"Prioritization strategy used to order transactions in the queue. S may be: gas_price - Prioritize txs with high gas price, consecutive - Prioritize consecutive transactions. Score such that the first transaction has a higher priority to reflect more transactions after it",
```
Comment (tomusdrw) [inline parity/cli/mod.rs]: I'd rephrase it a bit:
Include the entire chain of transactions when calculating their score. It means that one can increase priority of a transaction in the pool by sending next transaction after it.

File: miner/src/pool/scoring.rs (hunk 10 above)
Comment (insipx) [reply miner/src/pool/scoring.rs]: This is something I considered doing before but wasn't sure if it was necessary, plotting all the different algorithms now! :+1:

File: parity/cli/mod.rs (hunk 11 above)
Comment (insipx) [reply parity/cli/mod.rs]: changed it to `Include the entire chain of transactions when calculating their score. This means one may increase the priority of a transaction in the pool by sending another transaction from the same account`

File: miner/src/pool/scoring.rs (hunk 10 above)
Comment (tomusdrw) [reply miner/src/pool/scoring.rs]: Seems that the 3rd graph has a different scale.

Note that we onyl care about transactions above 21k (base gas), would be good to plot starting from there.
Do you mind plotting all options on the same chart and add a line for regular scoring as well (i.e. the one with gas price only?)

File: miner/src/pool/scoring.rs (hunk 9 above)
Comment (tomusdrw) [reply miner/src/pool/scoring.rs]: Still valid?

**Comment by 5chdn** [comment]
does not compile @insipx

File: miner/src/pool/scoring.rs (hunk 10 above)
Comment (insipx) [reply miner/src/pool/scoring.rs]: Merged all the graphs into two, one where Gas price was kept constant and another where Gas was kept at a constant. 


Legend for all the graphs:
Algorithm from PR:  Black,
Option 1: Blue
Option 2: Red
GasPriceOnly Scoring:  Green

## Constant Gas Price. Gas price kept at 10,000

Constant Gas Price

#### Same graph, but with an extended X-Axis
Constant Gas Price

## Constant Gas. Gas kept at 21,000
Constant Gas

**Comment by 5chdn** [comment]
Still seems to fail

**Comment by grbIzl** [comment]
@insipx as far as I've understood, the changes do not cover the attack with clearing all the balance (described in the original proposal). Am I correct? Do we have plans to address it later?

File: miner/src/pool/scoring.rs [hunk 12]
```diff
@@ -92,33 +92,67 @@ impl<P> txpool::Scoring<P> for NonceAndGasPrice where P: ScoredTransaction + txp
...
-				for (score, tx) in scores.iter_mut().zip(txs) {
-					// Never penalize local transactions.
-					if !tx.priority().is_local() {
-						*score = *score >> 3;
+			PrioritizationStrategy::Consecutive => {
+				let is_consecutive = |i: usize| {
+					if (txs[i].nonce() + U256::one()) == txs[i+1].nonce() {
+						U256::one()
+					} else {
+						U256::zero()
```
Comment (ngotchac) [inline miner/src/pool/scoring.rs]: Are the transactions sorted here? What happens if I send consecutive transactions in the wrong order?

File: miner/src/pool/scoring.rs [hunk 13]
```diff
@@ -92,33 +92,67 @@ impl<P> txpool::Scoring<P> for NonceAndGasPrice where P: ScoredTransaction + txp
...
-			// So just lower the priority of all non-local transactions.
-			Change::Event(_) => {
-				for (score, tx) in scores.iter_mut().zip(txs) {
-					// Never penalize local transactions.
-					if !tx.priority().is_local() {
-						*score = *score >> 3;
+			PrioritizationStrategy::Consecutive => {
+				let is_consecutive = |i: usize| {
+					if (txs[i].nonce() + U256::one()) == txs[i+1].nonce() {
+						U256::one()
```
Comment (ngotchac) [inline miner/src/pool/scoring.rs]: Why not directly return the `consecutive_bump` here?

File: miner/src/pool/scoring.rs (hunk 12 above)
Comment (insipx) [reply miner/src/pool/scoring.rs]: this part of the code should be dealing with sorted transactions. As I understand it, the sorting/ordering of transactions occurs in the `compare`, `choose` and `should_replace` methods as defined by the `Scoring` trait.

File: miner/src/pool/scoring.rs [hunk 14]
```diff
@@ -92,33 +92,67 @@ impl<P> txpool::Scoring<P> for NonceAndGasPrice where P: ScoredTransaction + txp
...
+					} else {
+						U256::zero()
 					}
+				};
+				match change {
+					Change::Culled(_) => {},
+					Change::RemovedAt(i) | Change::InsertedAt(i) | Change::ReplacedAt(i) => {
+						assert!(i < txs.len());
+						assert!(i < scores.len());
```
[the remark this answers is not in the document]
Comment (insipx) [reply miner/src/pool/scoring.rs]: Nope, it's not. Fixed that in my latest commit

**Comment by dvdplm** [comment]
@insipx I'm going to put this on ice for the time being. When you're ready for review again, just holler ok? :)

**Comment by debris** [comment]
closing as stale
